crates/nu-cli/src/shell/completer.rs

```diff
@@ -173,4 +209,72 @@ impl NuCompleter {
 
         matching_arguments
     }
+
+    // These is_executable/pathext implementations are copied from ichwh and modified
+    // to not be async
+
+    #[cfg(windows)]
+    fn pathext(&self) -> IchwhResult<Vec<String>> {
+        Ok(std::env::var_os("PATHEXT")
+            .ok_or(IchwhError::PathextNotDefined)?
+            .to_string_lossy()
+            .split(';')
+            // Cut off the leading '.' character
+            .map(|ext| ext[1..].to_string())
+            .collect::<Vec<_>>())
+    }
+
+    #[cfg(windows)]
+    fn is_executable(&self, file: &DirEntry) -> IchwhResult<bool> {
+        let file_type = file.metadata()?.file_type();
+
+        // If the entry isn't a file, it cannot be executable
+        if !(file_type.is_file() || file_type.is_symlink()) {
+            return Ok(false);
+        }
+
+        if let Some(extension) = file.path().extension() {
+            let exts = self.pathext()?;
+
+            Ok(exts
+                .iter()
+                .any(|ext| extension.to_string_lossy().eq_ignore_ascii_case(ext)))
+        } else {
+            Ok(false)
+        }
+    }
+
+    #[cfg(unix)]
+    fn is_executable(&self, file: &DirEntry) -> IchwhResult<bool> {
+        let metadata = file.metadata()?;
+
+        let filetype = metadata.file_type();
+        let permissions = metadata.permissions();
+
+        // The file is executable if it is a directory or a symlink and the permissions are set for
+        // owner, group, or other
+        Ok((filetype.is_file() || filetype.is_symlink()) && (permissions.mode() & 0o111 != 0))
+    }
+
+    fn get_path_executables(&self) -> Option<HashSet<String>> {
```

Comment:
Naming tip that I've learned over the years: `get_` is almost never a useful prefix. Functions that return something almost always "get" you something 🙂

Reply:
True! `find` works better here I think.